app/src/main/java/org/openobservatory/ooniprobe/activity/RunTestActivity.java

```diff
@@ -0,0 +1,265 @@
+                    }
+                    else {
+                        Alert.alertDialogWithAction(this, getString(R.string.invalid_parameter), getString(R.string.test_name) +  " : " + tn, new DialogInterface.OnClickListener() {
+                            public void onClick(DialogInterface dialog, int id) {
+                                goToMainActivity();
+                            }
+                        });
+                    }
+                }
+                else {
+                    Alert.alertDialogWithAction(this, getString(R.string.ooniprobe_outdate), getString(R.string.ooniprobe_outdate_msg),
+                            new DialogInterface.OnClickListener() {
+                                public void onClick(DialogInterface dialog, int id) {
+                                    final String appPackageName = getPackageName(); // getPackageName() from Context or Activity object
+                                    try {
+                                        openPlayStore("market://details?id=" + appPackageName);
+                                    } catch (android.content.ActivityNotFoundException anfe) {
+                                        openPlayStore("https://play.google.com/store/apps/details?id=" + appPackageName);
+                                    }
+                                }
+                        },
+                            new DialogInterface.OnClickListener() {
+                                public void onClick(DialogInterface dialog, int id) {
+                                    goToMainActivity();
+                                }
+                            });
+                }
+            }
+        }
+    }
+
+    public void configureScreen(String td, String ta){
+        TestData.getInstance(this, this).addObserver(this);
+
+        if (td != null)
+            title.setText(td);
+        else
+            title.setText(getString(R.string.run_test_message));
+
+        String test = NetworkMeasurement.getTestName(this, test_name);
+        testTitle.setText(test);
+        testImage.setImageResource(NetworkMeasurement.getTestImageBig(test_name));
+
+        ArrayList<String> listItems = new ArrayList<>();
+        final ArrayList<String> urlItems = new ArrayList<>();
+        if (ta != null){
+            try {
+                JSONObject taJson = new JSONObject(ta);
+                JSONArray urls = taJson.getJSONArray("urls");
+                for (int i = 0; i < urls.length(); i++)
+                    urlItems.add(urls.getString(i));
+            } catch (JSONException e) {
+                Alert.alertDialogWithAction(this, getString(R.string.invalid_parameter), getString(R.string.urls), new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int id) {
+                        goToMainActivity();
+                    }
+                });
+                e.printStackTrace();
```

Comment:
I don't think it's correct to continue after the exception. You should probably return here, no?

Reply:
I show an alert and close the activity as you requested

Reply:
Well, but then the code falls through and you call:

- `listItems.addAll(urlItems);`
- then there is a series of ifs that modify `mUrlListAdapter`
- then you call `runButton.OnClickListener`
- then ...

My point is: if, on error, you want to _stop_ the execution of the current function, you should `return`.

I am missing something obvious?

Reply:
Now I understand you, I call finish() only when the user clicks OK in the alert